File: src/main/java/blippy/metainfo/MetainfoDecoder.java
package blippy.metainfo;

import blippy.encoding.Bencoding;
import blippy.metainfo.Metainfo.Info;
import blippy.metainfo.Metainfo.MultipleFileInfo;
import blippy.metainfo.Metainfo.MultipleFileInfo.File;
import blippy.metainfo.Metainfo.SingleFileInfo;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;
import java.util.Map;

public class MetainfoDecoder {

  public MetainfoDecoder() {}

  @SuppressWarnings("unchecked")
  public Metainfo decode(final InputStream torrent) {
    try (torrent) {
      final Map<String, Object> metainfo = Bencoding.decodeDictionary(torrent.readAllBytes());
      final URI announce = new URI((String) metainfo.get("announce"));
      final Map<String, Object> infoDictionary = (Map<String, Object>) metainfo.get("info");
      final Info info;
      if (infoDictionary.containsKey("files")) {
        // Multiple File Mode
        final List<File> files =
            ((List<Map<String, Object>>) infoDictionary.get("files"))
                .stream()
                    .map(
                        fileDictionary -> {
                          final long length = (long) fileDictionary.get("length");
                          final List<String> path = (List<String>) fileDictionary.get("path");
                          return new File(length, path);
                        })
                    .toList();
        info = new MultipleFileInfo(infoDictionary, files);
      } else {
        // Single File Mode
        info = new SingleFileInfo(infoDictionary);
      }
      return new Metainfo(announce, info);
    } catch (final IOException | URISyntaxException ex) {
      throw new RuntimeException(ex);
    }
  }
}

File: src/main/kotlin/blippy/metainfo/MetainfoDecoder.kt
package blippy.metainfo

import blippy.crypto.Hashing
import blippy.encoding.Bencoding
import blippy.metainfo.Info.MultipleFileInfo
import blippy.metainfo.Info.SingleFileInfo
import java.io.IOException
import java.io.InputStream
import java.net.URI
import java.net.URISyntaxException

class MetainfoDecoder {

    @Suppress("UNCHECKED_CAST")
    fun decode(torrent: InputStream): Metainfo {
        try {
            torrent.use {
                val metainfo = Bencoding.decodeDictionary(torrent.readAllBytes())
                val announce = URI(metainfo["announce"] as String)
                val infoDictionary = metainfo["info"] as Map<String, Any>
                val infoHash = Hashing.sha1(Bencoding.encodeMap(infoDictionary))
                val info: Info = if (infoDictionary.containsKey("files")) {
                    // Multiple File Mode
                    val files =
                        (infoDictionary["files"] as List<Map<String, Any>>)
                            .stream()
                            .map { fileDictionary: Map<String, Any> ->
                                val length = fileDictionary["length"] as Long
                                val path = fileDictionary["path"] as List<String>
                                MultipleFileInfo.File(length, path)
                            }
                            .toList()
                    MultipleFileInfo(infoDictionary, infoHash, files)
                } else {
                    // Single File Mode
                    SingleFileInfo(infoDictionary, infoHash)
                }
                return Metainfo(announce, info)
            }
        } catch (ex: IOException) {
            throw RuntimeException(ex)
        } catch (ex: URISyntaxException) {
            throw RuntimeException(ex)
        }
    }
}
